### ISSUE TITLE
Naive idea for a solver/integrator

### ISSUE THREAD

**Issue by MuleaneEve** [bepu/bepuphysics2#55]
Let me start with: I have no clue how effective/realistic this idea is.
After reading your issue on substepping, I had this idea, and I thought it sounded cool :)

To illustrate the idea, let's take an example that is typically difficult to deal with in physics engines: Imagine that we fire two bullets that are supposed to collide.
![image](https://user-images.githubusercontent.com/32024816/48670412-d7acc780-eb17-11e8-89eb-af918e3e0dce.png)

Normally, considering the velocity and size of the bullets, we would need unrealistically tiny steps for the physics engine to correctly catch that collision, but what if the physics engine used the following high-level algorithm:

1. We imagine that each bullet is occupying all of the space between t and t+1 while following its current position and velocity.
![image](https://user-images.githubusercontent.com/32024816/48670415-e0050280-eb17-11e8-9728-ed17e78f60ed.png)

2. We figure out at exactly what instant "t+x" these two spaces collide.
3. We run the collision algo to figure out the new trajectories of these bullets
4. We start over (at 1.) with the time interval being between t+x and t+1.

Step 4 starts a recursive loop that is key because it enables handling a scenario where there is a 3rd bullet that one of the first two is supposed to collide with after its trajectory has been altered, but before t+1.
![image](https://user-images.githubusercontent.com/32024816/48670418-e85d3d80-eb17-11e8-816d-6b211bd7887c.png)

The benefit of this algo is that we have accurate collisions while being able to use big time steps.
Do you think there is any hope that it can be fast enough?
Or maybe it can still be useful in scenarios where accuracy is more important than performance (like when rendering a movie)?

**Comment by RossNordby** [comment]
This can indeed work, and it can indeed be fast enough for some use cases. For example, using a bouncing ray cast to simulate a laser beam is a pretty common special case of this idea. More generally, this can be thought of as a kind of adaptive step size where collision events are used as a heuristic.

In fact, @erincatto did something similar to this for dynamic-static collision in Diablo 3. You can read more in his 2013 presentation: http://box2d.org/files/GDC2013/ErinCatto_GDC2013.zip

Using a strict version of this for all dynamic interactions is, as you might expect, not very fast. Inserting a substep boundary for every interaction quickly becomes nasty when dealing with more than a few dynamic objects interacting at the same time. There are ways of mitigating this- substepping only the local potentially interacting subset of the simulation, minimum advancement thresholds, ignoring many collisions (like dynamic-dynamic or just any collision below a velocity threshold)- but it's always going to come at a cost.

Also, to get a fully 'correct' continuous result, the sweep test and integrator used for the substeps must include all relevant influences. For example, a ragdoll's body parts aren't going to follow perfectly linear or even ballistic trajectories due to all the interacting constraints. That can get expensive and it means piecewise linear sweep tests and integrators are still going to produce error in the general case. Definitely better than not doing anything, of course.

The fact that Bepuphysics v2 focuses on working with very large blobs of things also causes a bit of a mismatch. For example, constraints and convex collision pairs aren't actually evaluated one by one- they're executed in SIMD bundles and on multiple threads. Island local substepping and single event evaluation make that kind of machine friendly design more difficult.

(There is still hope for continuous collision detection even without any form of substepping, though. For example, speculative contacts can be used to punt the problem of multi-event collision response to the velocity solver in a pretty elegant way. It's got some tradeoffs and it won't often converge to the same result as a simulation running at 100000hz, but it works pretty darn well if you can generate appropriate contacts without causing ghost collisions.)

In summary, this kind of thing can definitely work, and it has been used. But it doesn't quite fit the aims of bepuphysics v2, so I wouldn't expect to see it show up.

**Comment by MuleaneEve** [comment]
Nice! I'm glad to see that this idea has already been explored.
And I like computer science because there are so many different ways to approach a problem.

One of bepuphysics v2 main attraction is its high-performance focus, so if you don't think it fits, then I'm fine with that.
